from sqlalchemy import create_engine, text

database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

connection_string

engine = create_engine(connection_string)

engine

#Initial EDA
query = '''
SELECT *
FROM scooters
LIMIT 10;
'''

with engine.connect() as connection:
    result = connection.execute(text(query))

result.fetchall()

In [1]:
import pandas as pd

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Bird';
'''

with engine.connect() as connection:
    bird = pd.read_sql(text(query), con = connection)

bird.head()

bird.to_csv("birddf.csv", index = False)

In [2]:
#read in bird
birddf = pd.read_csv('../data/birddf.csv')

In [3]:
birddf.shape

(12251590, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Bolt';
'''

with engine.connect() as connection:
    bolt = pd.read_sql(text(query), con = connection)

bolt.head()

bolt.to_csv("boltdf.csv", index = False)

In [4]:
#Read in bolt
boltdf = pd.read_csv('../data/boltdf.csv')

In [5]:
boltdf.shape

(3477198, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Gotcha';
'''

with engine.connect() as connection:
    gotcha = pd.read_sql(text(query), con = connection)

gotcha.head()

gotcha.to_csv("gotchadf.csv", index = False)

In [6]:
#Read in gotcha
gotchadf = pd.read_csv('../data/gotchadf.csv')

In [7]:
gotchadf.shape

(4679280, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Jump';
'''

with engine.connect() as connection:
    jump = pd.read_sql(text(query), con = connection)

jump.head()

#Saving jump to csv

jump.to_csv("jumpdf.csv", index = False)

In [8]:
#Read in jump
jumpdf = pd.read_csv('../data/jumpdf.csv')

In [9]:
jumpdf.shape

(21835098, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Lime';
'''

with engine.connect() as connection:
    lime = pd.read_sql(text(query), con = connection)

lime.head()

lime.to_csv("limedf.csv", index = False)

In [10]:
limedf = pd.read_csv('../data/limedf.csv')

In [11]:
limedf.shape

(16524261, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Lyft';
'''

with engine.connect() as connection:
    lyft = pd.read_sql(text(query), con = connection)

lyft.head()

lyft.to_csv("lyftdf.csv", index = False)

In [12]:
lyftdf = pd.read_csv('../data/lyftdf.csv')

In [13]:
lyftdf.shape

(9087043, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Spin';
'''

with engine.connect() as connection:
    spin = pd.read_sql(text(query), con = connection)

spin.head()

spin.to_csv("spindf.csv", index = False)

In [14]:
spindf = pd.read_csv('../data/spindf.csv')

In [15]:
spindf.shape

(5559573, 9)

query = '''
SELECT *
FROM trips;
'''

with engine.connect() as connection:
    tripsdf = pd.read_sql(text(query), con = connection)

tripsdf.head()

tripsdf.to_csv("tripsdf.csv", index = False)

query = '''
SELECT *
FROM trips;
'''

with engine.connect() as connection:
    trips = pd.read_sql(text(query), con = connection)

trips.head()

In [16]:
tripsdf = pd.read_csv('../data/tripsdf.csv')

In [17]:
tripsdf.shape

(565522, 16)

In [18]:
tripsdf.info

<bound method DataFrame.info of                    pubtimestamp companyname triprecordnum  \
0       2019-05-01 00:00:55.423        Bird       BRD2134   
1       2019-05-01 00:03:33.147        Lyft          LFT5   
2       2019-05-01 00:05:55.570        Bird       BRD2168   
3       2019-05-01 00:05:55.570        Bird       BRD2166   
4       2019-05-01 00:05:55.570        Bird       BRD2165   
...                         ...         ...           ...   
565517  2019-08-01 04:53:48.000        JUMP          JMP3   
565518  2019-08-01 04:53:48.000        JUMP          JMP2   
565519  2019-08-01 04:53:48.000        JUMP          JMP1   
565520  2019-08-01 07:04:00.000        JUMP         JMP35   
565521  2019-08-01 07:04:00.000        JUMP         JMP34   

                                             sumdid  tripduration  \
0                                      Powered9EAJL      3.000000   
1                                     Powered296631      1.715600   
2                           

In [19]:
#Concat to create the scooters table
scooters_list = [birddf, boltdf, gotchadf, jumpdf, limedf, lyftdf, spindf]
scootersdf = pd.concat(scooters_list)

In [20]:
scootersdf.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0  2019-07-30 17:37:28.870  36.141562 -86.806655  PoweredVAW9Y  Powered   
1  2019-07-30 17:37:28.870  36.159744 -86.780067  PoweredSNB79  Powered   
2  2019-07-30 17:37:28.870  36.153569 -86.784961  PoweredZY5KX  Powered   
3  2019-07-30 17:37:28.870  36.155984 -86.782756  PoweredPHVAG  Powered   
4  2019-07-30 17:37:28.870  36.147253 -86.815000  Powered238ZE  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         95.0   scooter        0.15        Bird  
1         58.0   scooter        0.15        Bird  
2        100.0   scooter        0.15        Bird  
3        100.0   scooter        0.15        Bird  
4         76.0   scooter        0.15        Bird

In [21]:
scootersdf.info()

<class 'pandas.core.frame.DataFrame'>
Index: 73414043 entries, 0 to 5559572
Data columns (total 9 columns):
 #   Column       Dtype  
---  ------       -----  
 0   pubdatetime  object 
 1   latitude     float64
 2   longitude    float64
 3   sumdid       object 
 4   sumdtype     object 
 5   chargelevel  float64
 6   sumdgroup    object 
 7   costpermin   float64
 8   companyname  object 
dtypes: float64(4), object(5)
memory usage: 5.5+ GB


## - Are there any null values in any columns in either table?

In [22]:
#Checking for NaN values in the scooters dataframe
scootersdf.isna().any()

pubdatetime    False
latitude       False
longitude      False
sumdid         False
sumdtype       False
chargelevel     True
sumdgroup      False
costpermin     False
companyname    False
dtype: bool

In [23]:
scootersdf.isnull().sum().sum()

770

In [24]:
#Checking for NaN values in the trips dataframe
tripsdf.isna().any()

pubtimestamp      False
companyname       False
triprecordnum     False
sumdid            False
tripduration      False
tripdistance      False
startdate         False
starttime         False
enddate           False
endtime           False
startlatitude     False
startlongitude    False
endlatitude       False
endlongitude      False
triproute         False
create_dt         False
dtype: bool

## - What date range is represented in each of the date columns? Investigate any values that seem odd.

In [25]:
scootersdf['pubdatetime'] = pd.to_datetime(scootersdf['pubdatetime'], dayfirst = False, format = 'mixed')

In [26]:
scootersdf.dtypes

pubdatetime    datetime64[ns]
latitude              float64
longitude             float64
sumdid                 object
sumdtype               object
chargelevel           float64
sumdgroup              object
costpermin            float64
companyname            object
dtype: object

In [27]:
scootersdf['pubdatetime'].max()

Timestamp('2019-07-31 23:59:57')

In [28]:
scootersdf['pubdatetime'].min()

Timestamp('2019-05-01 00:01:41.247000')

In [29]:
tripsdf['pubtimestamp'] = pd.to_datetime(tripsdf['pubtimestamp'], dayfirst = False, format = 'mixed')

In [30]:
tripsdf.dtypes

pubtimestamp      datetime64[ns]
companyname               object
triprecordnum             object
sumdid                    object
tripduration             float64
tripdistance             float64
startdate                 object
starttime                 object
enddate                   object
endtime                   object
startlatitude            float64
startlongitude           float64
endlatitude              float64
endlongitude             float64
triproute                 object
create_dt                 object
dtype: object

#Checking out the pubtimestamp column of the trips table
tripsdf["pubtimestamp"]

#Checking out the create_dt column in the trips df
tripsdf["create_dt"]

In [31]:
tripsdf['pubtimestamp'].min()

Timestamp('2019-05-01 00:00:55.423000')

In [32]:
tripsdf['pubtimestamp'].max()

Timestamp('2019-08-01 07:04:00')

The scooters df date range is '2019-05-01 00:01:41.247000' - '2019-07-31 23:59:57'.
The trips df date range is '2019-05-01 00:00:55.423000' - '2019-08-01 07:04:00'

## - Is time represented with am/pm or using 24 hour values in each of the columns that include time?

Time is represented in 24 hour values in the time columns in both dataframes.

## - What values are there in the sumdgroup column? Are there any that are not of interest for this project?

In [33]:
scootersdf.sumdgroup.unique()

array(['scooter', 'Scooter', 'bicycle'], dtype=object)

We don't need the bicycle sumdgroup for this project.

## - What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [34]:
scootersdf['latitude'].min()

0.0

In [35]:
scootersdf['latitude'].max()

3609874.116666

In [36]:
scootersdf['longitude'].min()

-97.443879

In [37]:
scootersdf['longitude'].max()

0.0

In [38]:
tripsdf['startlatitude'].min()

35.8532

In [39]:
tripsdf['startlatitude'].max()

36.300029

In [40]:
tripsdf['startlongitude'].min()

-86.918008

In [41]:
tripsdf['startlongitude'].max()

-86.3662

The minimum latitude shown is "0.0", as well as the maximum longitude, which is surprising. The maximum latitude is '3609874.116666' which is also "surprising" to say the least.

## -What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.

In [42]:
tripsdf['tripduration'].min()

-19.3582666667

In [43]:
tripsdf['tripduration'].max()

512619.0

In [44]:
tripsdf['tripdistance'].min()

-20324803.8

In [45]:
tripsdf['tripdistance'].max()

31884482.6476

In [46]:
tripsdf.query('`tripduration`== -19.3582666667')

pubtimestamp companyname triprecordnum         sumdid  \
509599 2019-07-19 00:14:02.297        Lyft         LFT21  Powered853770   

        tripduration  tripdistance   startdate        starttime     enddate  \
509599    -19.358267    4540.68256  2019-07-19  00:09:04.506666  2019-07-18   

                endtime  startlatitude  startlongitude  endlatitude  \
509599  23:49:43.013333       36.15505       -86.78482      36.1469   

        endlongitude                                          triproute  \
509599     -86.79355  [(36.15505, -86.78482), (36.15511, -86.78482),...   

                      create_dt  
509599  2019-07-20 10:52:39.737

In [47]:
tripsdf.query('`tripdistance`== -20324803.8')

pubtimestamp companyname triprecordnum        sumdid  \
550038 2019-07-27 23:56:41.940        Bird       BRD5324  PoweredYZL2A   

        tripduration  tripdistance   startdate        starttime     enddate  \
550038          44.0   -20324803.8  2019-07-27  23:14:23.683333  2019-07-27   

                endtime  startlatitude  startlongitude  endlatitude  \
550038  23:58:23.820000        36.1585        -86.7683       36.126   

        endlongitude                                          triproute  \
550038      -86.7892  [(36.158517, -86.768224), (36.158624, -86.7682...   

                      create_dt  
550038  2019-07-28 05:32:05.610

In [48]:
tripsdf.query('`tripdistance`== 31884482.6476')

pubtimestamp companyname triprecordnum       sumdid  \
447799 2019-07-05 01:18:52      Gotcha          GOT9  Powered2460   

        tripduration  tripdistance   startdate starttime     enddate  \
447799           9.0  3.188448e+07  2019-07-05  01:18:56  2019-07-05   

         endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
447799  01:27:56       36.16338      -86.780142     36.16628    -86.778127   

                                      triproute                create_dt  
447799  [('0', '0'), ('36.16628', '-86.77813')]  2019-07-06 07:13:12.210

## - Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?


In [49]:
scootersdf.companyname.unique()

array(['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin'],
      dtype=object)

In [50]:
tripsdf.companyname.unique()

array(['Bird', 'Lyft', 'Lime', 'Bolt Mobility', 'SPIN', 'Gotcha', 'JUMP'],
      dtype=object)

In the trips df, "Bolt" is "Bolt Mobility", "Jump", and "Spin" are also in upper case.

In [54]:
tripsdf["companyname"] = tripsdf["companyname"].replace("Bolt Mobility", "Bolt").replace("JUMP", "Jump").replace("SPIN", "Spin")

In [56]:
tripsdf.query('`companyname` == "Spin"')

pubtimestamp companyname triprecordnum          sumdid  \
34402  2019-06-14 21:39:10.037        Spin        SPI625  Powered1515735   
46545  2019-06-11 01:20:50.737        Spin         SPI57  Powered6743157   
48938  2019-06-26 01:28:48.173        Spin         SPI22  Powered8052471   
48958  2019-06-26 01:28:48.173        Spin         SPI25  Powered6183762   
49057  2019-06-26 01:28:48.173        Spin         SPI24  Powered7347865   
...                        ...         ...           ...             ...   
565322 2019-07-31 23:59:08.867        Spin        SPI191  Powered2403097   
565473 2019-08-01 00:09:09.333        Spin        SPI200  Powered5299331   
565479 2019-08-01 00:19:09.543        Spin        SPI199  Powered1086663   
565493 2019-08-01 00:54:10.327        Spin        SPI197  Powered3320563   
565514 2019-08-01 02:14:12.163        Spin        SPI201  Powered6219542   

        tripduration  tripdistance   startdate starttime     enddate  \
34402            9.0     928.47772  2019-06-14  21:30:16  2019-06-14   
46545            7.0    5075.45948  2019-06-11  01:09:56  2019-06-11   
48938           10.0    1955.38064  2019-06-26  01:14:51  2019-06-26   
48958            9.0    3989.50144  2019-06-26  01:15:55  2019-06-26   
49057           10.0    1781.49612  2019-06-26  01:15:23  2019-06-26   
...              ...           ...         ...       ...         ...   
565322          46.0     170.60368  2019-07-31  23:10:57  2019-07-31   
565473          17.0    4589.89516  2019-07-31  23:52:17  2019-08-01   
565479          24.0    1463.25464  2019-07-31  23:51:54  2019-08-01   
565493          71.0    2208.00532  2019-07-31  23:43:49  2019-08-01   
565514         140.0     239.50132  2019-07-31  23:54:20  2019-08-01   

         endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
34402   21:38:36      36.157994      -86.778604    36.155924    -86.780443   
46545   01:16:14      36.139189      -86.800526    36.152314    -86.794841   
48938   01:24:31      36.176276      -86.756405    36.176962    -86.749816   
48958   01:24:35      36.153228      -86.793661    36.158907    -86.782098   
49057   01:25:06      36.148626      -86.806844    36.151759    -86.802209   
...          ...            ...             ...          ...           ...   
565322  23:56:38      36.152557      -86.793373    36.152945    -86.793706   
565473  00:09:02      36.158799      -86.782449    36.171179    -86.785158   
565479  00:15:53      36.165682      -86.772403    36.164465    -86.777139   
565493  00:53:54      36.148838      -86.806735    36.149384    -86.814198   
565514  02:14:09      36.148476      -86.807011    36.147899    -86.807403   

                                                triproute  \
34402   [(36.15628, -86.78037), (36.1559240101242, -86...   
46545   [(36.152314230467226, -86.79484067953706), (36...   
48938   [(36.1769618, -86.749816), (36.17684, -86.7501...   
48958   [(36.15885, -86.78175), (36.1589075, -86.78209...   
49057   [(36.1517588299562, -86.80220881743135), (36.1...   
...                                                   ...   
565322  [(36.1529451, -86.7937055), (36.15282, -86.793...   
565473  [(36.1711785, -86.7851584), (36.17104, -86.785...   
565479  [(36.1644648760894, -86.77713928752279), (36.1...   
565493  [(36.1493, -86.81414), (36.1493, -86.81414), (...   
565514  [(36.14808, -86.80634), (36.14805, -86.80655),...   

                      create_dt  
34402   2019-06-15 09:28:22.217  
46545   2019-06-12 09:06:48.120  
48938   2019-06-27 10:21:51.893  
48958   2019-06-27 10:21:51.933  
49057   2019-06-27 10:21:51.980  
...                         ...  
565322  2019-08-01 11:36:29.957  
565473  2019-08-02 11:30:29.293  
565479  2019-08-02 11:30:29.327  
565493  2019-08-02 11:30:29.827  
565514  2019-08-02 11:30:29.923  

[34450 rows x 16 columns]

## 1. During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?

In [57]:
scootersdf.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0 2019-07-30 17:37:28.870  36.141562 -86.806655  PoweredVAW9Y  Powered   
1 2019-07-30 17:37:28.870  36.159744 -86.780067  PoweredSNB79  Powered   
2 2019-07-30 17:37:28.870  36.153569 -86.784961  PoweredZY5KX  Powered   
3 2019-07-30 17:37:28.870  36.155984 -86.782756  PoweredPHVAG  Powered   
4 2019-07-30 17:37:28.870  36.147253 -86.815000  Powered238ZE  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         95.0   scooter        0.15        Bird  
1         58.0   scooter        0.15        Bird  
2        100.0   scooter        0.15        Bird  
3        100.0   scooter        0.15        Bird  
4         76.0   scooter        0.15        Bird

In [58]:
scootersdf['Month'] = scootersdf['pubdatetime'].dt.month

In [59]:
scooters_may = scootersdf[scootersdf['Month'] == 5]

In [60]:
scooters_may.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
145127 2019-05-01 00:01:41.247  36.136822 -86.799877  PoweredLIRL1  Powered   
145128 2019-05-01 00:01:41.247  36.191252 -86.772945  PoweredXWRWC  Powered   
145129 2019-05-01 00:01:41.247  36.144752 -86.806293  PoweredMEJEH  Powered   
145130 2019-05-01 00:01:41.247  36.162056 -86.774688  Powered1A7TC  Powered   
145131 2019-05-01 00:01:41.247  36.150973 -86.783109  Powered2TYEF  Powered   

        chargelevel sumdgroup  costpermin companyname  Month  
145127         93.0   scooter         0.0        Bird      5  
145128         35.0   scooter         0.0        Bird      5  
145129         90.0   scooter         0.0        Bird      5  
145130         88.0   scooter         0.0        Bird      5  
145131         98.0   scooter         0.0        Bird      5

In [61]:
may_company_scooters = scooters_may[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'maycount')\
                                                        .sort_values(['maycount'])

In [62]:
may_company_scooters

companyname  maycount
2      Gotcha       223
1        Bolt       346
6        Spin       687
3        Jump       976
5        Lyft      1376
4        Lime      1632
0        Bird      3064

In [63]:
scooters_june = scootersdf[scootersdf['Month'] == 6]

In [64]:
june_company_scooters = scooters_june[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'junecount')\
                                                        .sort_values(['junecount'])

In [65]:
june_company_scooters

companyname  junecount
2      Gotcha        223
1        Bolt        333
6        Spin        713
3        Jump       1036
5        Lyft       1152
4        Lime       1461
0        Bird       2910

In [66]:
scooters_july = scootersdf[scootersdf['Month'] == 7]

In [67]:
july_company_scooters = scooters_july[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'julycount')\
                                                        .sort_values(['julycount'])

In [68]:
july_company_scooters

companyname  julycount
2      Gotcha        224
1        Bolt        276
6        Spin        570
4        Lime        903
5        Lyft       1026
3        Jump       1206
0        Bird       2583

In [69]:
may_june = may_company_scooters.merge(june_company_scooters, how = 'inner', on = 'companyname')

In [70]:
company_scooters_months = may_june.merge(july_company_scooters, how = 'inner', on = 'companyname')

In [71]:
company_scooters_months

companyname  maycount  junecount  julycount
0      Gotcha       223        223        224
1        Bolt       346        333        276
2        Spin       687        713        570
3        Jump       976       1036       1206
4        Lyft      1376       1152       1026
5        Lime      1632       1461        903
6        Bird      3064       2910       2583

In [72]:
tripsdf

pubtimestamp companyname triprecordnum  \
0      2019-05-01 00:00:55.423        Bird       BRD2134   
1      2019-05-01 00:03:33.147        Lyft          LFT5   
2      2019-05-01 00:05:55.570        Bird       BRD2168   
3      2019-05-01 00:05:55.570        Bird       BRD2166   
4      2019-05-01 00:05:55.570        Bird       BRD2165   
...                        ...         ...           ...   
565517 2019-08-01 04:53:48.000        Jump          JMP3   
565518 2019-08-01 04:53:48.000        Jump          JMP2   
565519 2019-08-01 04:53:48.000        Jump          JMP1   
565520 2019-08-01 07:04:00.000        Jump         JMP35   
565521 2019-08-01 07:04:00.000        Jump         JMP34   

                                             sumdid  tripduration  \
0                                      Powered9EAJL      3.000000   
1                                     Powered296631      1.715600   
2                                      Powered7S2UU      3.000000   
3                                      PoweredZIIVX      3.000000   
4                                      PoweredJ7MB3      2.000000   
...                                             ...           ...   
565517  Powered5614e4fc-1971-5e5c-a545-0fd88cd3331d     17.497950   
565518  Powereda4712099-5e55-5332-996c-d6e5c910535d      3.154917   
565519  Poweredb8a3a269-d1ca-571f-9b2f-89b7399b5537     18.440650   
565520  Poweredf077a919-d569-5e70-8ca7-71d179ffacf9    142.345610   
565521  Powered784fcc92-bd1a-5f6a-8314-b02aa21b4bfa    209.409650   

        tripdistance   startdate        starttime     enddate  \
0          958.00528  2019-05-01  00:00:20.460000  2019-05-01   
1         1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2         2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3         1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4          351.04988  2019-05-01  00:04:27.796666  2019-05-01   
...              ...         ...              ...         ...   
565517    7075.20000  2019-07-31         23:47:43  2019-08-01   
565518    1320.00000  2019-07-31         23:57:19  2019-08-01   
565519    7920.00000  2019-07-31         23:41:52  2019-08-01   
565520   20433.60000  2019-07-31         23:26:15  2019-08-01   
565521   14889.60000  2019-07-31         22:19:08  2019-08-01   

                endtime  startlatitude  startlongitude  endlatitude  \
0       00:02:52.346666      36.157100      -86.803600    36.156600   
1       00:03:33.026666      36.157970      -86.778960    36.160540   
2       00:07:13.596666      36.154700      -86.781800    36.156500   
3       00:06:59.176666      36.149400      -86.779500    36.153100   
4       00:06:23.150000      36.177800      -86.786600    36.177400   
...                 ...            ...             ...          ...   
565517         00:05:13      36.154633      -86.798340    36.149220   
565518         00:00:28      36.153015      -86.783585    36.152843   
565519         00:00:18      36.153687      -86.784580    36.164158   
565520         01:48:35      36.155735      -86.775185    36.173904   
565521         01:48:33      36.160050      -86.776720    36.163410   

        endlongitude                                          triproute  \
0         -86.806700  [(36.157235, -86.803612), (36.157235, -86.8036...   
1         -86.776890  [(36.15797, -86.77896), (36.15795, -86.77873),...   
2         -86.786800  [(36.155068, -86.782124), (36.156597, -86.78675)]   
3         -86.779600  [(36.149741, -86.779344), (36.149741, -86.7793...   
4         -86.787600  [(36.177699, -86.786477), (36.177711, -86.7864...   
...              ...                                                ...   
565517    -86.813980  [('36.154631', '-86.798341'), ('36.154631', '-...   
565518    -86.779594  [('36.153016', '-86.783586'), ('36.153016', '-...   
565519    -86.776900  [('36.153685', '-86.784578'), ('36.153685', '-...   
565520    -86.785450  [('36.155736', '-86.775181'), ('36.155531', '-...   
565521   

In [73]:
#Counting pubtimestap per company
tripsdf[['companyname', 'pubtimestamp']].groupby(['companyname'])['pubtimestamp']\
                                                        .nunique()

companyname
Bird       17242
Bolt        8072
Gotcha      2489
Jump         239
Lime       68852
Lyft      120925
Spin        9993
Name: pubtimestamp, dtype: int64

## 2. According to Second Substitute Bill BL2018-1202 (as amended) (https://web.archive.org/web/20181019234657/https://www.nashville.gov/Metro-Clerk/Legislative/Ordinances/Details/7d2cf076-b12c-4645-a118-b530577c5ee8/2015-2019/BL2018-1202.aspx), all permitted operators will first clean data before providing or reporting data to Metro. Data processing and cleaning shall include:  
* Removal of staff servicing and test trips  
* Removal of trips below one minute  
* Trip lengths are capped at 24 hours  
Are the scooter companies in compliance with the second and third part of this rule?

In [74]:
#Looking for trips below one minute.
short_trips = tripsdf.query('`tripduration` < 1')

In [75]:
short_trips[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values(['count'])

companyname  count
1        Lime    383
0        Bird   1229
2        Lyft   1329

In [76]:
#Looking for trips longer than 24 hours
long_trips = tripsdf.query('`tripduration` > 1440')

In [77]:
long_trips[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values(['count'])

companyname  count
1        Lyft      2
2        Spin     28
0        Bolt    351

## Answer: The following 5 companies are not in compliance with the 2nd/3rd parts of Bill BL2018-1202- Bird, Bolt, Lime, Lyft, and Spin, with Lyft failing to comply with either part. Bird, Lime, and Lyft failed to remove trips shorter than 1 minute. Bolt, Lyft, & Spin failed to remove trips longer than 24 hours.

## 3. The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. Based on the data, what is the average number of trips per scooter per day? Make sure to consider the days that a scooter was available. How does this vary by company?


From Postgres: Bird has 152,745 sumdids (3,766 distinct)
               Bolt has 21,890 sumdids (356 distinct)
               Gotcha has 3,315 sumdids (166 distinct)
               Jump has 6,437 sumdids (450 distinct)
               Lime has 225,694 sumdids (1,788 distinct)
               Lyft has 120,991 sumdids (1,725 distinct)
               Spin has 34,450 sumdids (754 distinct)

In [78]:
sumdid_per_day = tripsdf.groupby(['sumdid', 'companyname', 'startdate']).size()\
                        .reset_index(name='times_used_per_day')

In [79]:
sumdid_per_day['times_used_per_day'].mean()

2.793225395382837

In [80]:
scooters_trips_per_day = sumdid_per_day.groupby('sumdid')['times_used_per_day'].sum().reset_index()

In [81]:
scooters_trips_per_day

sumdid  times_used_per_day
0     Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619                  92
1     Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d                  57
2     Powered-03be23ca-d43b-222f-be54-e44b5b4690df                  81
3     Powered-046201fb-6532-1f37-6334-3612fb1e61f7                  71
4     Powered-0479bb84-afbd-0426-f1c4-df628542a88c                  96
...                                            ...                 ...
9000   Poweredff356f94-2252-5432-b484-ad9679bf4b8d                  11
9001   Poweredff4b01f3-1f3f-5d6b-b1c5-5d6628ffa629                   6
9002   Poweredff8a28ab-0ffb-59ce-9b1e-2d1299a25634                  23
9003   Poweredffcc1b11-af95-5e06-b002-82ffb342aae8                   6
9004                         StandardLIRPOZ7X3ZVGO                   1

[9005 rows x 2 columns]

In [82]:
scooters_used_per_co = sumdid_per_day.groupby('companyname')['times_used_per_day'].size().reset_index()

In [83]:
scooters_used_per_co

companyname  times_used_per_day
0        Bird               77386
1        Bolt                9509
2      Gotcha                1357
3        Jump                3024
4        Lime               54878
5        Lyft               39841
6        Spin               16467

scooters_pings_per_date['pings_per_date'].mean()

tripsdf.pivot(index='sumdid', columns='companyname', values='pubtimestamp')

In [84]:
scootersdf['Date'] = scootersdf['pubdatetime'].dt.day

In [85]:
scootersdf

pubdatetime   latitude  longitude          sumdid  \
0       2019-07-30 17:37:28.870  36.141562 -86.806655    PoweredVAW9Y   
1       2019-07-30 17:37:28.870  36.159744 -86.780067    PoweredSNB79   
2       2019-07-30 17:37:28.870  36.153569 -86.784961    PoweredZY5KX   
3       2019-07-30 17:37:28.870  36.155984 -86.782756    PoweredPHVAG   
4       2019-07-30 17:37:28.870  36.147253 -86.815000    Powered238ZE   
...                         ...        ...        ...             ...   
5559568 2019-07-30 17:30:06.803  36.153090 -86.783980  Powered8052471   
5559569 2019-07-30 17:30:06.803  36.158640 -86.767830  Powered4853276   
5559570 2019-07-30 17:30:06.803  36.152500 -86.783750  Powered1245835   
5559571 2019-07-30 17:30:06.803  36.158020 -86.783790  Powered7927186   
5559572 2019-07-30 17:30:06.803  36.164380 -86.769250  Powered4053444   

        sumdtype  chargelevel sumdgroup  costpermin companyname  Month  Date  
0        Powered         95.0   scooter        0.15        Bird      7    30  
1        Powered         58.0   scooter        0.15        Bird      7    30  
2        Powered        100.0   scooter        0.15        Bird      7    30  
3        Powered        100.0   scooter        0.15        Bird      7    30  
4        Powered         76.0   scooter        0.15        Bird      7    30  
...          ...          ...       ...         ...         ...    ...   ...  
5559568  Powered         92.0   Scooter        0.23        Spin      7    30  
5559569  Powered         99.0   Scooter        0.23        Spin      7    30  
5559570  Powered         96.0   Scooter        0.23        Spin      7    30  
5559571  Powered        100.0   Scooter        0.23        Spin      7    30  
5559572  Powered         99.0   Scooter        0.23        Spin      7    30  

[73414043 rows x 11 columns]

In [86]:
scooters_pings_per_date = scootersdf.groupby(['sumdid', 'companyname', 'Month', 'Date']).size()\
                                    .reset_index(name='trips_per_day')

In [87]:
#Double checking the count of a particular scooter
scooters_pings_per_date[scooters_pings_per_date['sumdid']=='Powered5129907'].Date.count()

21

#Checking answers and getting a count that is too high
scootersdf[scootersdf['sumdid']=='Powered5129907'].drop_duplicates().groupby(['sumdid', 'companyname'])['Date'].count()\
                                    .reset_index(name='count')

#Looking at values for one particular scooter
scootersdf[scootersdf['sumdid']== "Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619"]

In [88]:
scooters_pings_per_date

sumdid companyname  Month  Date  \
0       Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt      5    24   
1       Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt      5    25   
2       Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt      5    26   
3       Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt      5    27   
4       Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt      5    28   
...                                              ...         ...    ...   ...   
413558                         StandardNW5HJFO4R32LY        Lime      7    28   
413559                         StandardNW5HJFO4R32LY        Lime      7    29   
413560                         StandardNW5HJFO4R32LY        Lime      7    30   
413561                         StandardNW5HJFO4R32LY        Lime      7    31   
413562                         StandardZPUQESHVPP74J        Lime      6    13   

        trips_per_day  
0                 181  
1                 263  
2                  62  
3                 149  
4                  84  
...               ...  
413558            288  
413559            288  
413560            287  
413561            288  
413562              6  

[413563 rows x 5 columns]

In [89]:
scooters_pings_per_date.groupby('sumdid')['Date'].count().reset_index()

sumdid  Date
0      Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619    68
1      Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d    47
2      Powered-03be23ca-d43b-222f-be54-e44b5b4690df    62
3      Powered-046201fb-6532-1f37-6334-3612fb1e61f7    59
4      Powered-0479bb84-afbd-0426-f1c4-df628542a88c    66
...                                             ...   ...
10013                         Standard5JXOV277MCWID     1
10014                         StandardNPOOZNUSGAXZN     1
10015                         StandardNUTLLXP4G37OI     1
10016                         StandardNW5HJFO4R32LY    92
10017                         StandardZPUQESHVPP74J     1

[10018 rows x 2 columns]

In [90]:
scooters_avail_per_co = scooters_pings_per_date.groupby('companyname')['Date'].count().reset_index(name = 'avail_per_day')

In [91]:
scooters_avail_per_co

companyname  avail_per_day
0        Bird         134639
1        Bolt          16474
2      Gotcha          16762
3        Jump          77724
4        Lime          77265
5        Lyft          56230
6        Spin          34469

In [92]:
scooters_avail_use_co = scooters_used_per_co.merge(scooters_avail_per_co, how = 'outer', on = 'companyname')

In [93]:
scooters_avail_use_co

companyname  times_used_per_day  avail_per_day
0        Bird               77386         134639
1        Bolt                9509          16474
2      Gotcha                1357          16762
3        Jump                3024          77724
4        Lime               54878          77265
5        Lyft               39841          56230
6        Spin               16467          34469

tripsdf["sumdid"].value_counts()

tripsdf[['power_type', 'sumdid']] = tripsdf['sumdid'].str.split(' ', expand=True)

trips_days = (tripsdf
 .assign(day = tripsdf['pubtimestamp'].dt.day_name())
 .day
 .value_counts(sort = False)
)

trips_days

(tripsdf
 .groupby(pd.Grouper(key = 'pubtimestamp',     
                     freq = '1d',               
                     origin = 'epoch'          
                                               
                    ))
 ['sumdid']
 .count()
 .reset_index()
 .head(10)
)

(tripsdf
 .assign(day_name = tripsdf['startdate'].dt.day_name())
 .groupby('day_name')
 ['sumdid']
 .nunique()
)

In [94]:
tripsdf[['companyname', 'sumdid', 'startdate']].groupby(['companyname', 'startdate'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values(['count'])

companyname   startdate  count
214      Gotcha  2019-07-23      1
218        Jump  2019-05-20      1
219        Jump  2019-06-10      1
220        Jump  2019-06-11      1
221        Jump  2019-06-12      1
..          ...         ...    ...
39         Bird  2019-06-09   1410
25         Bird  2019-05-26   1483
38         Bird  2019-06-08   1513
58         Bird  2019-06-30   1546
57         Bird  2019-06-29   1566

[533 rows x 3 columns]

In [95]:
scootersdf.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0 2019-07-30 17:37:28.870  36.141562 -86.806655  PoweredVAW9Y  Powered   
1 2019-07-30 17:37:28.870  36.159744 -86.780067  PoweredSNB79  Powered   
2 2019-07-30 17:37:28.870  36.153569 -86.784961  PoweredZY5KX  Powered   
3 2019-07-30 17:37:28.870  36.155984 -86.782756  PoweredPHVAG  Powered   
4 2019-07-30 17:37:28.870  36.147253 -86.815000  Powered238ZE  Powered   

   chargelevel sumdgroup  costpermin companyname  Month  Date  
0         95.0   scooter        0.15        Bird      7    30  
1         58.0   scooter        0.15        Bird      7    30  
2        100.0   scooter        0.15        Bird      7    30  
3        100.0   scooter        0.15        Bird      7    30  
4         76.0   scooter        0.15        Bird      7    30

In [96]:
tripsdf.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance   startdate        starttime     enddate  \
0        3.0000     958.00528  2019-05-01  00:00:20.460000  2019-05-01   
1        1.7156    1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2        3.0000    2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3        3.0000    1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4        2.0000     351.04988  2019-05-01  00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute                create_dt  
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780  
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757  
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530  
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237  
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207

## 4. Metro would like to know how many scooters are needed, and something that could help with this is knowing peak demand. Estimate the highest count of scooters being used at the same time. When were the highest volume times? Does this vary by zip code or other geographic region?

In [97]:
scootersdf['Time'] = scootersdf['pubdatetime'].dt.time

In [98]:
scootersdf

pubdatetime   latitude  longitude          sumdid  \
0       2019-07-30 17:37:28.870  36.141562 -86.806655    PoweredVAW9Y   
1       2019-07-30 17:37:28.870  36.159744 -86.780067    PoweredSNB79   
2       2019-07-30 17:37:28.870  36.153569 -86.784961    PoweredZY5KX   
3       2019-07-30 17:37:28.870  36.155984 -86.782756    PoweredPHVAG   
4       2019-07-30 17:37:28.870  36.147253 -86.815000    Powered238ZE   
...                         ...        ...        ...             ...   
5559568 2019-07-30 17:30:06.803  36.153090 -86.783980  Powered8052471   
5559569 2019-07-30 17:30:06.803  36.158640 -86.767830  Powered4853276   
5559570 2019-07-30 17:30:06.803  36.152500 -86.783750  Powered1245835   
5559571 2019-07-30 17:30:06.803  36.158020 -86.783790  Powered7927186   
5559572 2019-07-30 17:30:06.803  36.164380 -86.769250  Powered4053444   

        sumdtype  chargelevel sumdgroup  costpermin companyname  Month  Date  \
0        Powered         95.0   scooter        0.15        Bird      7    30   
1        Powered         58.0   scooter        0.15        Bird      7    30   
2        Powered        100.0   scooter        0.15        Bird      7    30   
3        Powered        100.0   scooter        0.15        Bird      7    30   
4        Powered         76.0   scooter        0.15        Bird      7    30   
...          ...          ...       ...         ...         ...    ...   ...   
5559568  Powered         92.0   Scooter        0.23        Spin      7    30   
5559569  Powered         99.0   Scooter        0.23        Spin      7    30   
5559570  Powered         96.0   Scooter        0.23        Spin      7    30   
5559571  Powered        100.0   Scooter        0.23        Spin      7    30   
5559572  Powered         99.0   Scooter        0.23        Spin      7    30   

                    Time  
0        17:37:28.870000  
1        17:37:28.870000  
2        17:37:28.870000  
3        17:37:28.870000  
4        17:37:28.870000  
...                  ...  
5559568  17:30:06.803000  
5559569  17:30:06.803000  
5559570  17:30:06.803000  
5559571  17:30:06.803000  
5559572  17:30:06.803000  

[73414043 rows x 12 columns]

In [99]:
#Attempting to count number of scooters available at each time interval but the numbers aren't what I'm expecting. Needs tweaking.
scootersdf[['companyname', 'sumdid', 'Time']].groupby(['Time'])['Time']\
                                                        .count()\
                                                        .reset_index(name = 'scooters_count')\
                                                        .sort_values(['Time'])

Time  scooters_count
0               00:00:00             390
1        00:00:00.010000               1
2        00:00:00.017000               1
3        00:00:00.020000               1
4        00:00:00.033000               1
...                  ...             ...
7665002  23:59:59.947000               1
7665003  23:59:59.957000               1
7665004  23:59:59.970000               1
7665005  23:59:59.983000               3
7665006  23:59:59.993000               1

[7665007 rows x 2 columns]

#Counting number of trips per time interval
peak_time = tripsdf[['companyname', 'sumdid', 'starttime']].groupby(['starttime'])['starttime']\
                                                        .count()\
                                                        .reset_index(name = 'trips_count')\
                                                        .sort_values(['trips_count'], ascending=False)
peak_time.head(10)

In [100]:
tripsdf['startdatetime'] = pd.to_datetime(tripsdf['startdate'].astype('str')+' '+tripsdf['starttime'].astype('str'), format ='mixed')

In [101]:
tripsdf.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance   startdate        starttime     enddate  \
0        3.0000     958.00528  2019-05-01  00:00:20.460000  2019-05-01   
1        1.7156    1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2        3.0000    2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3        3.0000    1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4        2.0000     351.04988  2019-05-01  00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute                create_dt  \
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780   
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757   
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530   
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237   
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207   

               startdatetime  
0 2019-05-01 00:00:20.460000  
1 2019-05-01 00:01:50.090000  
2 2019-05-01 00:03:47.363333  
3 2019-05-01 00:04:21.386666  
4 2019-05-01 00:04:27.796666

In [102]:
(tripsdf
 .assign(hour = tripsdf['startdatetime'].dt.hour)     
 .groupby(['hour'])
 ['startdatetime']
 .count()
 .reset_index()
)

hour  startdatetime
0      0          43462
1      1          32690
2      2          24463
3      3          21827
4      4          16280
5      5           9471
6      6           6601
7      7           4334
8      8           2194
9      9           1132
10    10           1524
11    11           3630
12    12           7413
13    13          11705
14    14          17846
15    15          25907
16    16          36433
17    17          43339
18    18          44717
19    19          44698
20    20          41973
21    21          40902
22    22          38149
23    23          44832

In [103]:
tripsdf

pubtimestamp companyname triprecordnum  \
0      2019-05-01 00:00:55.423        Bird       BRD2134   
1      2019-05-01 00:03:33.147        Lyft          LFT5   
2      2019-05-01 00:05:55.570        Bird       BRD2168   
3      2019-05-01 00:05:55.570        Bird       BRD2166   
4      2019-05-01 00:05:55.570        Bird       BRD2165   
...                        ...         ...           ...   
565517 2019-08-01 04:53:48.000        Jump          JMP3   
565518 2019-08-01 04:53:48.000        Jump          JMP2   
565519 2019-08-01 04:53:48.000        Jump          JMP1   
565520 2019-08-01 07:04:00.000        Jump         JMP35   
565521 2019-08-01 07:04:00.000        Jump         JMP34   

                                             sumdid  tripduration  \
0                                      Powered9EAJL      3.000000   
1                                     Powered296631      1.715600   
2                                      Powered7S2UU      3.000000   
3                                      PoweredZIIVX      3.000000   
4                                      PoweredJ7MB3      2.000000   
...                                             ...           ...   
565517  Powered5614e4fc-1971-5e5c-a545-0fd88cd3331d     17.497950   
565518  Powereda4712099-5e55-5332-996c-d6e5c910535d      3.154917   
565519  Poweredb8a3a269-d1ca-571f-9b2f-89b7399b5537     18.440650   
565520  Poweredf077a919-d569-5e70-8ca7-71d179ffacf9    142.345610   
565521  Powered784fcc92-bd1a-5f6a-8314-b02aa21b4bfa    209.409650   

        tripdistance   startdate        starttime     enddate  \
0          958.00528  2019-05-01  00:00:20.460000  2019-05-01   
1         1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2         2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3         1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4          351.04988  2019-05-01  00:04:27.796666  2019-05-01   
...              ...         ...              ...         ...   
565517    7075.20000  2019-07-31         23:47:43  2019-08-01   
565518    1320.00000  2019-07-31         23:57:19  2019-08-01   
565519    7920.00000  2019-07-31         23:41:52  2019-08-01   
565520   20433.60000  2019-07-31         23:26:15  2019-08-01   
565521   14889.60000  2019-07-31         22:19:08  2019-08-01   

                endtime  startlatitude  startlongitude  endlatitude  \
0       00:02:52.346666      36.157100      -86.803600    36.156600   
1       00:03:33.026666      36.157970      -86.778960    36.160540   
2       00:07:13.596666      36.154700      -86.781800    36.156500   
3       00:06:59.176666      36.149400      -86.779500    36.153100   
4       00:06:23.150000      36.177800      -86.786600    36.177400   
...                 ...            ...             ...          ...   
565517         00:05:13      36.154633      -86.798340    36.149220   
565518         00:00:28      36.153015      -86.783585    36.152843   
565519         00:00:18      36.153687      -86.784580    36.164158   
565520         01:48:35      36.155735      -86.775185    36.173904   
565521         01:48:33      36.160050      -86.776720    36.163410   

        endlongitude                                          triproute  \
0         -86.806700  [(36.157235, -86.803612), (36.157235, -86.8036...   
1         -86.776890  [(36.15797, -86.77896), (36.15795, -86.77873),...   
2         -86.786800  [(36.155068, -86.782124), (36.156597, -86.78675)]   
3         -86.779600  [(36.149741, -86.779344), (36.149741, -86.7793...   
4         -86.787600  [(36.177699, -86.786477), (36.177711, -86.7864...   
...              ...                                                ...   
565517    -86.813980  [('36.154631', '-86.798341'), ('36.154631', '-...   
565518    -86.779594  [('36.153016', '-86.783586'), ('36.153016', '-...   
565519    -86.776900  [('36.153685', '-86.784578'), ('36.153685', '-...   
565520    -86.785450  [('36.155736', '-86.775181'), ('36.155531', '-...   
565521   

In [104]:
import geopandas as gpd

## 5. **Stretch Goal:** SUMDs can provide alternative transportation and provide "last mile" access to public transit. How often are trips starting near public transit hubs? You can download a dataset of bus stop locations from https://data.nashville.gov/Transportation/WeGo-Transit-Bus-Stops/vfe9-k7vc/about_data.

In [105]:
busstopsdf = pd.read_csv('../data/busstopsdf.csv')
busstopsdf

Stop ID Stop Abbreviation                             Stop Name  \
0        5120          7AVUNINN                 7TH AVE & UNION ST NB   
1        5455             CM5_4                       Temporary Bay 4   
2        2762          23ACLASF      23RD AVE N & CLARKSVILLE PIKE SB   
3        4079          21ACHUSN             21ST AVE N & CHURCH ST SB   
4        4133           GRWHIWN       GREEN LN & WHITES CREEK PIKE WB   
...       ...               ...                                   ...   
2573     5166          JEF6AVWM             JEFFERSON ST & 6TH AVE WB   
2574     4698          NEEJAPEF     NEELY'S BEND RD & JASPERSON DR EB   
2575     4700          NEEALLEN       NEELY'S BEND RD & ALLEN PASS EB   
2576     4714          OLDFORWN  OLD HICKORY BLVD & FOREST PARK RD WB   
2577     1940          THOEIREF            THOMPSON LN & EAST IRIS DR   

      Shelter?  Bench? Route Numbers                 Route Name  \
0        False   False           NaN  AIRPORT - DOWNTOWN HOTELS   
1        False   False           NaN                     HERMAN   
2        False   False            22                   BORDEAUX   
3        False    True            25                    MIDTOWN   
4        False   False            41              GOLDEN VALLEY   
...        ...     ...           ...                        ...   
2573     False   False            60    MUSIC CITY BLUE CIRCUIT   
2574     False   False            76          MADISON CONNECTOR   
2575     False   False           NaN          MADISON CONNECTOR   
2576     False   False            76          MADISON CONNECTOR   
2577     False   False            77         THOMPSON CONNECTOR   

              Mapped Location  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4     (36.236249, -86.816722)  
...                       ...  
2573  (36.173796, -86.789456)  
2574  (36.247943, -86.682633)  
2575                      NaN  
2576  (36.263003, -86.703336)  
2577  (36.111799, -86.762652)  

[2578 rows x 8 columns]

In [106]:
trips_start_loc = tripsdf[["triprecordnum", "startlatitude"]]
trips_start_loc

triprecordnum  startlatitude
0            BRD2134      36.157100
1               LFT5      36.157970
2            BRD2168      36.154700
3            BRD2166      36.149400
4            BRD2165      36.177800
...              ...            ...
565517          JMP3      36.154633
565518          JMP2      36.153015
565519          JMP1      36.153687
565520         JMP35      36.155735
565521         JMP34      36.160050

[565522 rows x 2 columns]

In [107]:
busstops_loc = busstopsdf[["Stop ID", "Stop Name", "Mapped Location"]]
busstops_loc

Stop ID                             Stop Name          Mapped Location
0        5120                 7TH AVE & UNION ST NB                      NaN
1        5455                       Temporary Bay 4                      NaN
2        2762      23RD AVE N & CLARKSVILLE PIKE SB                      NaN
3        4079             21ST AVE N & CHURCH ST SB                      NaN
4        4133       GREEN LN & WHITES CREEK PIKE WB  (36.236249, -86.816722)
...       ...                                   ...                      ...
2573     5166             JEFFERSON ST & 6TH AVE WB  (36.173796, -86.789456)
2574     4698     NEELY'S BEND RD & JASPERSON DR EB  (36.247943, -86.682633)
2575     4700       NEELY'S BEND RD & ALLEN PASS EB                      NaN
2576     4714  OLD HICKORY BLVD & FOREST PARK RD WB  (36.263003, -86.703336)
2577     1940            THOMPSON LN & EAST IRIS DR  (36.111799, -86.762652)

[2578 rows x 3 columns]

In [108]:
busstops_loc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2578 entries, 0 to 2577
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Stop ID          2578 non-null   int64 
 1   Stop Name        2578 non-null   object
 2   Mapped Location  2524 non-null   object
dtypes: int64(1), object(2)
memory usage: 60.5+ KB


In [109]:
busstops_locs = busstops_loc.dropna()
busstops_locs

Stop ID                             Stop Name          Mapped Location
4        4133       GREEN LN & WHITES CREEK PIKE WB  (36.236249, -86.816722)
5        4083         _ 9TH AVE S & EDGEHILL AVE SB  (36.142642, -86.780897)
6        5066        DONELSON/DELL STATION OUTBOUND  (36.105615, -86.672004)
7          94            17TH AVE S & DOROTHY PL SB  (36.137623, -86.795609)
8        3219              COCKRILL ST & 14TH AVE N  (36.175944, -86.804242)
...       ...                                   ...                      ...
2572     5320         EZELL PIKE CONVENIENCE CTR WB  (36.086851, -86.686667)
2573     5166             JEFFERSON ST & 6TH AVE WB  (36.173796, -86.789456)
2574     4698     NEELY'S BEND RD & JASPERSON DR EB  (36.247943, -86.682633)
2576     4714  OLD HICKORY BLVD & FOREST PARK RD WB  (36.263003, -86.703336)
2577     1940            THOMPSON LN & EAST IRIS DR  (36.111799, -86.762652)

[2524 rows x 3 columns]

In [102]:
#Attempting to split the mapped location column but not getting expected results
busstops_locs[["Latitude", "Longitude"]] = busstops_locs["Mapped Location"].str.split(", ", expand=True)
#pd.DataFrame(busstops_locs["Mapped Location"].str.split(', ',1).tolist(), 
 #                        columns = ['Latitude','Longitude'])

C:\Users\adaly\AppData\Local\Temp\ipykernel_17948\3096984761.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  busstops_locs[["Latitude", "Longitude"]] = busstops_locs["Mapped Location"].str.split(", ", expand=True)
C:\Users\adaly\AppData\Local\Temp\ipykernel_17948\3096984761.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  busstops_locs[["Latitude", "Longitude"]] = busstops_locs["Mapped Location"].str.split(", ", expand=True)


In [103]:
busstops_locs

Stop ID                             Stop Name          Mapped Location  \
4        4133       GREEN LN & WHITES CREEK PIKE WB  (36.236249, -86.816722)   
5        4083         _ 9TH AVE S & EDGEHILL AVE SB  (36.142642, -86.780897)   
6        5066        DONELSON/DELL STATION OUTBOUND  (36.105615, -86.672004)   
7          94            17TH AVE S & DOROTHY PL SB  (36.137623, -86.795609)   
8        3219              COCKRILL ST & 14TH AVE N  (36.175944, -86.804242)   
...       ...                                   ...                      ...   
2572     5320         EZELL PIKE CONVENIENCE CTR WB  (36.086851, -86.686667)   
2573     5166             JEFFERSON ST & 6TH AVE WB  (36.173796, -86.789456)   
2574     4698     NEELY'S BEND RD & JASPERSON DR EB  (36.247943, -86.682633)   
2576     4714  OLD HICKORY BLVD & FOREST PARK RD WB  (36.263003, -86.703336)   
2577     1940            THOMPSON LN & EAST IRIS DR  (36.111799, -86.762652)   

        Latitude    Longitude  
4     (36.236249  -86.816722)  
5     (36.142642  -86.780897)  
6     (36.105615  -86.672004)  
7     (36.137623  -86.795609)  
8     (36.175944  -86.804242)  
...          ...          ...  
2572  (36.086851  -86.686667)  
2573  (36.173796  -86.789456)  
2574  (36.247943  -86.682633)  
2576  (36.263003  -86.703336)  
2577  (36.111799  -86.762652)  

[2524 rows x 5 columns]

In [ ]:
#Need to merge trips_loc and busstops_locs


## Deliverables:
At the conclusion of this project, your group should deliver a presentation which addresses the following points:
* Are scooter companies in compliance with the required data cleaning?
* What are typical usage patterns for scooters in terms of time, location, and trip duration?
* What are your recommendations for total number of scooters for the city overall and density of scooters by zip code?
* **Stretch Goal:** Does it appear that scooters are used as "last mile" transportation from public transit hubs to work or school?


-Only Gotcha and Jump are in compliance with the required data cleaning.

-

-

-